purlcli.py

```diff
@@ -0,0 +1,253 @@
+
+
+@purlcli.command(name="meta")
+@click.option(
+    "--purl",
+    "purls",
+    multiple=True,
+    required=False,
+    help="PackageURL or PURL.",
+)
+@click.option(
+    "--output",
+    type=click.File(mode="w", encoding="utf-8"),
+    required=True,
+    default="-",
+    help="Write versions output as JSON to FILE.",
+)
+@click.option(
+    "--file",
+    type=click.File(mode="r", encoding="utf-8"),
+    required=False,
+    help="Read a list of PURLs from a FILE, one per line.",
+)
+def get_meta(purls, output, file):
+    """
+    Given one or more PURLs, return a mapping of metadata fetched from the API for each PURL.
+    """
+    if purls and file:
+        raise click.UsageError("Use either purls or file but not both.")
+
+    if not (purls or file):
+        raise click.UsageError("Use either purls or file.")
+
+    if file:
+        purls = file.read().splitlines(False)
+
+    meta_info = get_meta_details(purls)
+    json.dump(meta_info, output, indent=4)
+
+
+def get_meta_details(purls):
+    meta_details = []
+    purl_versions = []
+
+    for purl in purls:
+        dict_output = {}
+        dict_output["purl"] = purl
+        dict_output["metadata"] = []
+
+        purl = purl.strip()
+        if not purl:
+            continue
+
+        if check_meta_purl(purl):
+            print(check_meta_purl(purl))
+            continue
+
+        releases = []
+        for release in list(info(purl)):
+            dict_output["metadata"].append(release.to_dict())
+
+        purl_versions.append(dict_output)
+
+    return purl_versions
+
+
+def check_existence(purl):
+    api_query = "https://public.purldb.io/api/validate/"
+    purl = purl.strip()
+    request_body = {"purl": purl, "check_existence": True}
+    response = requests.get(api_query, params=request_body)
+    results = response.json()
+
+    return results
+
+
+def check_versions_purl(purl):
+    results = check_existence(purl)
+
+    if results["valid"] == False:
+        return f"There was an error with your '{purl}' query -- the Package URL you provided is not valid."
+
+    supported = SUPPORTED_ECOSYSTEMS
+
+    purl_type = purl.split(":")[1].split("/")[0]
+
+    if purl_type not in supported:
+        return f"The provided PackageURL '{purl}' is valid, but `versions` is not supported for this package type."
+
+    if results["exists"] != True:
+        return f"There was an error with your '{purl}' query.  Make sure that '{purl}' actually exists in the relevant repository."
+
+
+def check_meta_purl(purl):
+    results = check_existence(purl)
+    if results["valid"] == False:
+        return f"There was an error with your '{purl}' query -- the Package URL you provided is not valid."
+
+    # This is manually constructed from a visual inspection of fetchcode/package.py:
+    SUPPORTED_ECOSYSTEMS = [
+        "cargo",
+        "npm",
+        "pypi",
+        "github",
+        "bitbucket",
+        "rubygems",
+    ]
+    purl_type = purl.split(":")[1].split("/")[0]
```

Comment:
I would create a `PackageURL` object from `purl` string, and then access the `type` field of that object

```suggestion
    purl = PackageURL.from_string(purl)
    if purl.type not in SUPPORTED_ECOSYSTEMS:
```

Reply:
@JonoYang Talk about over-engineering.  I was so proud of devising `purl.split(":")[1].split("/")[0]`, and overlooked the straightforward approach that you identified and that I've now incorporated.  Thank you.  ;-)  Reran my purlcli tests, all 42 passed.

Reran `make test` and ... error -- not sure what this is telling me to do:

```
. . .

_________________________________________________________________________________________________ ERROR collecting purldb-toolkit/setup.py _________________________________________________________________________________________________
import file mismatch:
imported module 'setup' has this __file__ attribute:
  /home/jmh/dev/nexb/purldb/setup.py
which is not the same as the test file we want to collect:
  /home/jmh/dev/nexb/purldb/purldb-toolkit/setup.py
HINT: remove __pycache__ / .pyc files and/or use a unique basename for your test file modules
========================================================================================================= short test summary info ==========================================================================================================
ERROR purldb-toolkit/setup.py
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
============================================================================================================= 1 error in 3.11s =============================================================================================================
make: *** [Makefile:110: test] Error 2

(venv) Thu Jan 18, 2024 12:27 PM  /home/jmh/dev/nexb/purldb jmh (247-create-purl-cli-tool)
$
```

Reply:
@johnmhoran 

The test setup is a bit convoluted. Pytest didn't play nice with the directory layout I created, where we have modules like `matchcode-toolkit` and `purldb-toolkit`. To stop that error, I had to tell pytest to ignore those directories, then I added in extra commands to pytest to just test those specific directories. 

We want to update this line: https://github.com/nexB/purldb/blob/247-create-purl-cli-tool/Makefile#L110

to `	${ACTIVATE} DJANGO_SETTINGS_MODULE=purldb_project.settings ${PYTHON_EXE} -m pytest -vvs --ignore matchcode-toolkit --ignore purldb-toolkit --ignore packagedb/tests/test_throttling.py`

and then add another line to test the `purldb-toolkit` specifically
`	${ACTIVATE} ${PYTHON_EXE} -m pytest -vvs purldb-toolkit`

Reply:
@JonoYang 

Did the pytest fine-tuning, reran 'make test'.  Only failures were the same 2 as the other day, which we expect atm.

```
FAILED minecode/tests/test_maven.py::MavenEnd2EndTest::test_visit_and_map_with_index - AssertionError: Lists differ: [{'ur[31 chars]ven2/cnuernber/dtype-next/0.4.2/dtype-next-0.4[49087 chars]one}] != [{'ur[31 chars]ven2/.index/nexus-mav
en-repository-index.532.g[49087 chars]one}]
FAILED minecode/tests/test_ls.py::ParseDirectoryListingTest::test_parse_listing_from_lslr - AssertionError: Lists differ: [{'pa[1527 chars] '2023-01', 'target': None}, {'path': 'dists/e[974 chars]one}] != [{'pa[1527 chars] '2024-01', 't
arget': None}, {'path': 'dists/e[974 chars]one}]

. . .

2 failed, 583 passed, 2 skipped, 30 xfailed, 71 warnings in 81.68s (0:01:21)

. . .
```

and all 42 of my purlcli tests pass.  Ready to commit and push.